fix TCP transport for web

lukediamond opened the pull request:
Makes sure the correct transport is chosen for web and node, adds shims for `net`/`tls`.

@misterupkeep

legastero committed:
Minor JXT cleanup for XEP-0030 and XEP-0084

legastero committed:
Make it possible to enforce types for JXT definition fields

legastero committed:
Merge branch 'master' of github.com:legastero/stanza

legastero committed:
Allow inspecting for result completion of RSM pager

legastero committed:
12.17.2

Commit:
add TCP transport

Commit:
fix TCP transport for web

misterupkeep commented on scripts/build.ts:
```diff
@@ -19,7 +24,12 @@ fileReplace('dist/es/Constants.js', '__STANZAJS_VERSION__', Pkg.version);
 Child('npm run compile:rollup');
 
-Child('mkdir dist/npm');
+if (!FS.existsSync("dist")) {
+    FS.mkdirSync("dist");
+}
+if (!FS.existsSync("dist/npm")) {
+    FS.mkdirSync("dist/npm");
+}
```
Why, though? The rest of the script file relies on gnutils anyways, might as well write it as `mkdir -p dist/npm`

lukediamond replied:
Because I'm running windows right now and I would rather replace this entire file with `fs` calls than get that set up and make it a dependency. It's completely unnecessary to call the shell here.

misterupkeep commented on src/Client.ts:
```diff
@@ -284,7 +288,8 @@ export default class Client extends EventEmitter {
                 bosh: true,
-                websocket: true
+                websocket: true,
+                tcp: false,
```
TCP should also be enabled by default
```suggestion
                tcp: true,
```

misterupkeep reviewed (changes requested):
TCP needs to be enabled by default, and I have to test this in both browser and node environments before merging.

misterupkeep committed:
fix(tcp): enable tcp by default

lukediamond commented:
Can we merge this?

misterupkeep commented:
I still have to test the actual TCP transport. I know it works (or rather, does _not_ work) in the browser, but haven't so much as tried connecting over TCP. 

Perhaps hooking up TCP into Electron builds would be the best, most adequate test, but _that_ is waiting on bonfire-xmpp/bonfire#60, which is in turn waiting on upstream.

Merging without testing (or taking your word) might be pushing unstable code to what should be a stable library. We can merge but wait on the release until we get TCP working on Bonfire. Even if it's broken, it should be working sans TCP transport.

jensengar committed:
bump async to fix browser throttling

legastero committed:
Merge pull request #491 from jensengar:bump-async

bump async to fix browser throttling

Commit:
Improve TCP host resolution

Commit:
Update live-connection test suite for TCP transport

Commit:
Improve transport selection

Commit:
Improve TCP testing

lukediamond committed:
Merge branch 'legastero:master' into tcp-transport

legastero committed:
Merge branch 'master' of github.com:legastero/stanza

Commit:
remove console.log calls

Commit:
Merge branch 'tcp-transport' of https://github.com/00drip/stanza into tcp-transport

legastero committed:
Stanza shims updated

legastero committed:
Better support for adding new transport types

- Can specify order to try transports using `transportPreferenceOrder`
  config.

- Can do binding discovery per transport type by implementing a
  `discoverBindings()` method on the transport object.

legastero committed:
Linting

legastero committed:
Fix compile errors from latest TypeScript

Commit:
merge upstream transport interface

Commit:
do not use unsupported transports

Commit:
restructure transports and add requireSecureTransport

Commit:
remove commented error code
